BREAKING: Remove virtual on methods that are being called from constructors

laimis opened the pull request:
Continuation of fixes with virtual calls being made from constructors. The issue originally reported by SonarCloud scans: https://sonarcloud.io/project/issues?resolved=false&rules=csharpsquid%3AS1699&id=apache_lucenenet and referenced in this issue: https://github.com/apache/lucenenet/issues/670

This time I focused on cases where a class is public, we are breaking the API, BUT, the method that was marked virtual was not calling any other virtual methods, was not being overridden in Lucene repo, and/or looked like a method that was not meant to be overridden to change the behavior of the class. Similarly to #798 , preferring explicit/clear/no surprises break in API over introducing locking/lazy initialization/additional allocations when the public consumers have a way to safely extend the class behavior.

This should eliminate the majority of "easy" changes, leaving us with a few "hairer" cases that we will tackle separately.

laimis commented:
Taking a look at the failures, somehow locally I did not trigger them...

NightOwl888 reviewed (changes requested):
This is going to take a few rounds of reviews to deal with. I left a few comments on the first 3 to give you some background.

Many of these were intended for extension by end users. For example a `Clear()` method could be implemented by a subclass to clear more than what the base class clears. But if you take a look at all collection types in .NET, none of their members are virtual. Of course, it is a PITA because it means you basically have to write collections from scratch instead of simply extending them.

NightOwl888 commented on src/Lucene.Net.Suggest/Spell/SpellChecker.cs:
```diff
@@ -142,8 +142,9 @@ public SpellChecker(Directory spellIndex, IStringDistance sd, IComparer<SuggestW
-        public virtual void SetSpellIndex(Directory spellIndexDir)
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        public void SetSpellIndex(Directory spellIndexDir)
```
I was thinking this one is a good candidate for `LazyInitializer`. Then I saw the comment that was added by the Lucene team:

```c#
// TODO: we should make this final as it is called in the constructor
```

I checked and as of Lucene 9.5.0 they still haven't changed that. But being that they intend to, I think we can keep this change.

NightOwl888 commented on src/Lucene.Net.Queries/Function/ValueSourceScorer.cs:
```diff
@@ -48,7 +48,9 @@ protected internal ValueSourceScorer(IndexReader reader, FunctionValues values)
-        public virtual void SetCheckDeletes(bool checkDeletes)
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        public void SetCheckDeletes(bool checkDeletes)
```
Let's move this logic to a private `SetCheckDeletesInternal()` method that is called by the constructor and the virtual `SetCheckDeletes()` method. This just needs some docs to tell overriders that their overridden logic won't run in the constructor, so they will need to set this flag appropriately in their own constructor (it's fine to suggest adding a private method that is called in both places).

laimis replied:
Made the fix here: https://github.com/apache/lucenenet/pull/805

laimis commented on src/Lucene.Net.Analysis.Common/Analysis/Util/OpenStringBuilder.cs:
```diff
@@ -56,7 +56,9 @@ public virtual int Length
-        public virtual void Set(char[] arr, int end)
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        public void Set(char[] arr, int end)
```
Fixed here https://github.com/apache/lucenenet/pull/803

laimis commented on src/Lucene.Net.Analysis.Common/Tartarus/Snowball/SnowballProgram.cs:
```diff
@@ -63,7 +63,9 @@ protected SnowballProgram()
-        public virtual void SetCurrent(string value)
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        public void SetCurrent(string value)
```
Fixed here: https://github.com/apache/lucenenet/pull/802

NightOwl888 commented on src/Lucene.Net/Util/SentinelIntSet.cs:
```diff
@@ -89,7 +89,9 @@ public SentinelInt32Set(int size, int emptyVal)
-        public virtual void Clear()
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        public void Clear()
```
This one can be made into private `ClearInternal()` that is called by the constructor and this virtual `Clear()` method.

laimis replied:
Fixed with https://github.com/apache/lucenenet/pull/806

NightOwl888 replied:
BTW - Let's add some documentation on `Clear()` letting users know that if they override the method they would also need to add the same logic into the constructor to match Lucene.

NightOwl888 commented on src/Lucene.Net/Search/DisjunctionMaxQuery.cs:
```diff
@@ -92,7 +92,9 @@ public DisjunctionMaxQuery(ICollection<Query> disjuncts, float tieBreakerMultipl
-        public virtual void Add(Query query)
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        public void Add(Query query)
```
This one can be made into private `AddInternal()` called by the constructor and virtual `Add()`.

laimis replied:
Fixed with https://github.com/apache/lucenenet/pull/807

NightOwl888 replied:
BTW - Let's add some documentation on `Add()` letting users know that if they override the method they would also need to add the same logic into the constructor to match Lucene.

NightOwl888 commented on src/Lucene.Net.Spatial/Prefix/Tree/Cell.cs:
```diff
@@ -175,7 +175,9 @@ private void B_fixLeaf()
-        public virtual void SetLeaf()
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        public void SetLeaf()
```
This one can be made into private `SetLeafInternal()` called by `B_fixLeaf()` and virtual `SetLeaf()`. In the subclass constructor, the user has access to both `m_leaf` and `IsLeaf` so they can read or set the value on construction after the base class. Public `SetLeaf()` calls can then run whatever business logic that happens when it is set.

laimis replied:
Fixed with https://github.com/apache/lucenenet/pull/808

NightOwl888 replied:
BTW - Let's add some documentation on `SetLeaf()` letting users know that if they override the method they would also need to add the same logic into the constructor to match Lucene.

laimis commented on src/Lucene.Net.Facet/Taxonomy/Directory/DirectoryTaxonomyReader.cs:
```diff
@@ -241,15 +241,19 @@ protected override TaxonomyReader DoOpenIfChanged()
-        protected virtual DirectoryReader OpenIndexReader(Directory directory)
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        protected DirectoryReader OpenIndexReader(Directory directory)
```
OK, I will leave this one alone here and will revisit it separately.

laimis commented:
Thank you for the review. I went ahead and addressed the issues with separate PRs. I left out DirectoryTaxonomyReader.cs and will take it in another PR. Closing this one without merging.

NightOwl888 commented on src/Lucene.Net.QueryParser/Flexible/Standard/Nodes/NumericRangeQueryNode.cs:
```diff
@@ -87,7 +87,9 @@ private static NumericType GetNumericDataType(J2N.Numerics.Number number)
-        public virtual void SetBounds(NumericQueryNode lower, NumericQueryNode upper,
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        public void SetBounds(NumericQueryNode lower, NumericQueryNode upper,
```
This one can be replaced by a private `SetBoundsInternal()` method called by the constructor and public virtual `SetBounds()`.

BTW - Let's add some documentation on `SetBounds()` letting users know that if they override the method they would also need to add the same logic into the constructor to match Lucene.

laimis replied:
Fixed in https://github.com/apache/lucenenet/pull/812

NightOwl888 commented on src/Lucene.Net/Store/ByteArrayDataOutput.cs:
```diff
@@ -51,12 +51,16 @@ public ByteArrayDataOutput()
-        public virtual void Reset(byte[] bytes)
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        public void Reset(byte[] bytes)
```
This can be changed to `ResetInternal()` called by both the constructor and the virtual `Reset()` methods.

BTW - Let's add some documentation on `Reset()` letting users know that if they override the method they would also need to add the same logic into the constructor to match Lucene.

Also, let's add some null guard clauses for `bytes` that throw `ArgumentNullException`.

laimis replied:
Fixed in https://github.com/apache/lucenenet/pull/814

NightOwl888 commented on src/Lucene.Net/Search/Sort.cs:
```diff
@@ -139,13 +139,17 @@ public Sort(params SortField[] fields)
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        public void SetSort(SortField field)
         {
```
These can be made into `SetSortInternal()` called by the constructor and virtual `SetSort()`.

BTW - Let's add some documentation on both overloads of `SetSort()` letting users know that if they override the method they would also need to add the same logic into the constructor to match Lucene.

laimis replied:
Fixed in https://github.com/apache/lucenenet/pull/813

NightOwl888 commented on src/Lucene.Net/Search/IndexSearcher.cs:
```diff
@@ -161,7 +161,9 @@ public IndexSearcher(IndexReaderContext context)
-        protected virtual LeafSlice[] Slices(IList<AtomicReaderContext> leaves)
+        // LUCENENET specific - S1699 - marked non-virtual because calling
+        // virtual members from the constructor is not a safe operation in .NET
+        protected LeafSlice[] Slices(IList<AtomicReaderContext> leaves)
```
This one can be made into `GetSlicesInternal()` called by the constructor and virtual `GetSlices()` method. Note adding a verb here is intentional to indicate this is an action. And since it is protected, this break won't affect many users. Just be sure to add a note that we renamed this from Lucene.

BTW - Let's add some documentation on `GetSlices()` letting users know that if they override the method they would also need to add the same logic into the constructor to match Lucene.

laimis replied:
Fixed in https://github.com/apache/lucenenet/pull/818

NightOwl888 reviewed (commented):
I added a few more review comments.

laimis commented on src/Lucene.Net.Highlighter/VectorHighlight/FieldPhraseList.cs:
Fixed in https://github.com/apache/lucenenet/pull/811
